BtConnectorLib/btconnectorlib2/src/main/java/org/thaliproject/p2p/btconnectorlib/BTConnector.java

```diff
@@ -96,80 +86,95 @@ public WifiBtStatus Start(String peerIdentifier, String peerName) {
+            // the HW is not supporting all needed stuff
             print_line("", "BT available: " + ret.isBtOk + ", wifi available: " + ret.isWifiOk);
             setState(State.NotInitialized);
-        } else if (ret.isBtEnabled && ret.isWifiEnabled) {
-             print_line("", "All stuf available and enabled");
-             startAll();
-        }else{
-            //we wait untill both Wifi & BT are turned on
+            return ret;
+        }
+
+        if (!ret.isBtEnabled  || !ret.isWifiEnabled) {
+            //we will be waiting untill both Wifi & BT are turned on
             setState(State.WaitingStateChange);
+            return ret;
         }
 
+        //all is good, so lets get started
+        print_line("", "All stuff available and enabled");
+        startAll();
         return ret;
     }
 
     public void Stop() {
         stopAll();
-        if (mWifiBase != null) {
-            WifiBase tmp = mWifiBase;
-            mWifiBase = null;
+        WifiBase tmp = mWifiBase;
+        mWifiBase = null;
+        if (tmp != null) {
             tmp.Stop();
         }
 
-        if (mBluetoothBase != null) {
-            BluetoothBase tmpb =mBluetoothBase;
-            mBluetoothBase = null;
+        BluetoothBase tmpb =mBluetoothBase;
+        mBluetoothBase = null;
+        if (tmpb != null) {
             tmpb.Stop();
         }
     }
 
     public enum TryConnectReturnValues{
         Connecting,
-        AlreadyAttemptingToConnect,
+        AlreadyAttemptingToConnect, // this test is wrong, don't want to change for compatibility
         NoSelectedDevice,
         BTDeviceFetchFailed
     }
 
     public TryConnectReturnValues TryConnect(ServiceItem selectedDevice) {
 
-        TryConnectReturnValues ret = TryConnectReturnValues.Connecting;
-        if(selectedDevice != null) {
-            if (mBTConnector_BtConnection != null) {
-                BluetoothDevice device = mBluetoothBase.getRemoteDevice(selectedDevice.peerAddress);
-                if (device != null) {
-                    // actually the ret will now be always true, since mBTConnector_BtConnection only checks if device is non-null
-                    mBTConnector_BtConnection.TryConnect(device, this.ConSettings.MY_UUID, selectedDevice.peerId, selectedDevice.peerName, selectedDevice.peerAddress);
-                }else{
-                    ret = TryConnectReturnValues.BTDeviceFetchFailed;
-                }
-            }else{
-                ret = TryConnectReturnValues.AlreadyAttemptingToConnect;
-            }
-        }else{
-            ret = TryConnectReturnValues.NoSelectedDevice;
+        if(selectedDevice == null) {
+            return TryConnectReturnValues.NoSelectedDevice;
         }
-        return ret;
+
+        BluetoothBase tmoBase = mBluetoothBase;
+        if(tmoBase == null) {// should never happen, would indicate uninitialized system
+            return TryConnectReturnValues.BTDeviceFetchFailed;
```

Comment:
I believe you are saying that someone called tryConnect without a successful call to start. In that case shouldn't we throw an exception? Otherwise how can the dev tell the difference between having screwed up and called TryConnect without start and having called for the wrong peerAddress (the next usage of this same response code)? This really should be, I believe, an exception.